groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -765,787 +1431,10708 @@ void Optional_DataImp<TYPE>::emplace(
+                const bsl::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator==(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+    // If bool(x) != bool(y), false; otherwise if bool(x) == false, true;
+    // otherwise *x == *y. Note that this function will fail to compile if
+    // 'LHS_TYPE' and 'RHS_TYPE' are not compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator!=(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator!=(const std::optional<LHS_TYPE>& lhs,
+                const bsl::optional<RHS_TYPE>& rhs);
+    // If bool(x) != bool(y), true; otherwise, if bool(x) == false, false;
+    // otherwise *x != *y.  Note that this function will fail to compile if
+    // 'LHS_TYPE' and 'RHS_TYPE' are not compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<(const bsl::optional<LHS_TYPE>& lhs,
+               const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<(const std::optional<LHS_TYPE>& lhs,
+               const bsl::optional<RHS_TYPE>& rhs);
+    // If !y, false; otherwise, if !x, true; otherwise *x < *y.  Note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>(const bsl::optional<LHS_TYPE>& lhs,
+               const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>(const std::optional<LHS_TYPE>& lhs,
+               const bsl::optional<RHS_TYPE>& rhs);
+    // If !x, false; otherwise, if !y, true; otherwise *x > *y. note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<=(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator<=(const std::optional<LHS_TYPE>& lhs,
+                const bsl::optional<RHS_TYPE>& rhs);
+    // If !x, true; otherwise, if !y, false; otherwise *x <= *y. Note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>=(const bsl::optional<LHS_TYPE>& lhs,
+                const std::optional<RHS_TYPE>& rhs);
+template <class LHS_TYPE, class RHS_TYPE>
+bool operator>=(const std::optional<LHS_TYPE>& lhs,
+                const bsl::optional<RHS_TYPE>& rhs);
+    // If !y, true; otherwise, if !x, false; otherwise *x >= *y. Note that this
+    // function will fail to compile if 'LHS_TYPE' and 'RHS_TYPE' are not
+    // compatible.
+
+#endif  // BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+
+template <class TYPE>
+BSLS_KEYWORD_CONSTEXPR bsl::optional<typename bsl::decay<TYPE>::type>
```

Comment:
it will not compile either way, 'bsl::string*' is not an AA type and will not have the constructor overloads that take an allocator. 
In general, functions and arrays will be stored in the non AA optional which is capable of handling both. Having 'allocate_optional' use decay means that it is symmetrical to 'make_optional' and that it will not compile if the deduced type is not allocator aware.
I think we want the decay there. We want the same type deduced here as we get with make_optional.

Reply:
Ok